raven_aiohttp.py

```diff
@@ -58,50 +78,138 @@ def keepalive(self):
+                        loop=self._loop
+                    )
+
+                    assert len(self._tasks) == 0
+        except asyncio.TimeoutError:
+            pass
+        finally:
+            if self.keepalive:
+                yield from self._client_session.close()
+
+    @asyncio.coroutine
+    def _worker(self):
+        while True:
             try:
-                resp = yield from asyncio.wait_for(
-                    session.post(url,
-                                 data=data,
-                                 compress=False,
-                                 headers=headers),
-                    self.timeout,
-                    loop=self._loop)
-                yield from resp.release()
+                data = yield from self._queue.get()
+
+                if data is ...:
+                    self._queue.put_nowait(...)
+                    break
+
+                url, data, headers, success_cb, failure_cb = data
+
+                yield from self._do_send(url, data, headers, success_cb,
+                                         failure_cb)
+            finally:
+                self._queue.task_done()
+
+    @asyncio.coroutine
+    def _do_send(self, url, data, headers, success_cb, failure_cb):
+        if self.keepalive:
+            session = self._client_session
+        else:
+            session = self._client_session_factory()
+
+        resp = None
+        try:
+            with async_timeout.timeout(self.timeout, loop=self._loop):
+                # timeout=None disables built-in aiohttp timeout
+                resp = yield from session.post(url, data=data, compress=False,
+                                               headers=headers, timeout=None)
+
                 code = resp.status
                 if code != 200:
                     msg = resp.headers.get('x-sentry-error')
                     if code == 429:
                         try:
-                            retry_after = int(resp.headers.get('retry-after'))
+                            retry_after = resp.headers.get('retry-after')
+                            retry_after = int(retry_after)
                         except (ValueError, TypeError):
                             retry_after = 0
                         failure_cb(RateLimited(msg, retry_after))
                     else:
                         failure_cb(APIError(msg, code))
                 else:
                     success_cb()
-            except Exception as exc:
+        except asyncio.CancelledError:
+            # do not mute asyncio.CancelledError
+            raise
+        except Exception as exc:
+            failure_cb(exc)
+        finally:
+            if resp is not None:
+                resp.release()
+            if not self.keepalive:
+                yield from session.close()
+
+    def async_send(self, url, data, headers, success_cb, failure_cb):
+        if self._closing:
+            raise RuntimeError('AioHttpTransport is closing')
```

Comment:
@asvetlov lets place there `failure_cb(RuntimeError('AioHttpTransport is closing'))` ?

Reply:
Yes please